### PR TITLE
Add alternative parameterization to negative binomial distribution #4126

### PR DESCRIPTION
**Pull request by ricardoV94**
Changes allow the specification of the NegativeBinomial distribution in terms of p (the probability of observing a success in each trial), and n (the target number of successes). This parametrization gives the likelihood for y, the number of failures observed until a target number of successes is reached.

I changed the docstrings and distribution initialization in line with other distribution that have multiple parametrizations.

I didn't figure out yet how the tests work, so I did not implement it (hence the WIP)! This should be straightforward, however, since the scipy negativebinomial distribution is implemented in terms of n and p. Any guidance in how to proceed is much appreciated!

Pinging @twiecki 

Ref #4126

### CODE REVIEW

**Comment by twiecki** [comment]
This is great -- thanks! Definitely tests are required, do you need help figuring that out?

**Comment by ricardoV94** [comment]
Yeah, I would need some help with that. I didn't find anything on the developer's guide about testing, and the test code is a bit opaque for me.

**Comment by AlexAndorra** [comment]
Thanks for the PR @ricardoV94 ! For the tests, I think you can take inspiration from the tests of Negative Binomial done with the current parametrization. Doing them with the new one should already be a good sample of use-cases

**Comment by ricardoV94** [comment]
@twiecki @AlexAndorra would the added test be enough?

**Comment by MarcoGorelli** [comment]
IMO there should also be a test which hits the various error messages you've added, e.g. when you set both `alpha` and `n` - is that something you'd be happy to add? Feel free to ping if you want/need help

**Comment by ricardoV94** [comment]
@MarcoGorelli That makes sense.

I am not sure how to do it. Could you direct me to a snippet where this type of test is implemented? For example, the Beta distribution also raises this type of ValueErrors, but I couldn't find if it is being tested anywhere.

**Comment by MarcoGorelli** [comment]
They do something similar in `pymc3/tests/test_starting.py`:

```python
    with raises(ValueError, match=r"Some variables not in the model: \['x2', 'y2'\]"):
        starting.allinmodel([x2, y2], model1)
    with raises(ValueError, match=r"Some variables not in the model: \['x2'\]"):
        starting.allinmodel([x2, y1], model1)
    with raises(ValueError, match=r"Some variables not in the model: \['x2'\]"):
        starting.allinmodel([x2], model1)
```

where `raises` was imported from `pytest` (see https://docs.pytest.org/en/stable/assert.html for more on this)

**Comment by ricardoV94** [comment]
@MarcoGorelli Thanks for the input. I added some tests that cover the new ValueError exceptions. What do you think?

**Comment by MarcoGorelli** [comment]
@ricardoV94 nice work!

I think this can be simplified a bit using `pytest.mark.paramterize`, something like this (I've only written the first two test cases, but it should be easy to extend):
```python
    @pytest.mark.parametrize(
        "mu, n, alpha, expected",
        [
            (5, None, None, "Incompatible parametrization. Must specify either alpha or n."),
            (None, 2, 2, "Incompatible parametrization Can't specify both alpha and n."),
            # other two test cases go here
        ]
    )
    def test_negative_binomial_init_fail(self, mu, n, alpha, expected):
        with Model():
            with pytest.raises(ValueError, match=expected):
                x = NegativeBinomial("x", mu=mu, n=n, alpha=alpha)
```

File: pymc3/distributions/discrete.py [hunk 1]
```diff
@@ -598,20 +598,60 @@ def NegBinom(a, m, x):
...
+        if alpha is None:
+            if n is not None:
+                alpha = n
+            else:
+                raise ValueError(
+                    "Incompatible parametrization. Must specify either alpha or n."
+                )
+        elif n is not None:
+            raise ValueError(
+                "Incompatible parametrization Can't specify both alpha and n."
```
Comment (MarcoGorelli) [inline pymc3/distributions/discrete.py]: missing full-stop?
```suggestion
                "Incompatible parametrization. Can't specify both alpha and n."
```

**Comment by ricardoV94** [comment]
@MarcoGorelli Thanks for the suggestion. I refactored the code with the `pytest.mark.parametrize` as you suggested. 

I also added a couple of tests for the different permutations of missing/over-specified parameters, do you think it is too much?

**Review by twiecki — APPROVED** [review]

**Comment by twiecki** [comment]
Thanks @ricardoV94! Is this your first PR?

**Comment by ricardoV94** [comment]
You're welcome. Yes it was the first one for me :)

**Comment by twiecki** [comment]
Congratulations then! :)

**Pull request merged**